query.go

```diff
@@ -84,174 +162,212 @@ func (dht *IpfsDHT) runDisjointQueries(ctx context.Context, d int, target string
 
 	// distribute the shuffled K closest peers as seeds among the "d" disjoint queries
 	for i := 0; i < len(seedPeers); i++ {
-		queries[i%d].localPeers.Add(seedPeers[i])
+		queries[i%d].seedPeers = append(queries[i%d].seedPeers, seedPeers[i])
 	}
 
 	// start the "d"  disjoint queries
+	queryDone := make(chan struct{}, d)
 	for i := 0; i < d; i++ {
 		query := queries[i]
 		go func() {
-			strictParallelismQuery(query)
+			query.runWithGreedyParallelism()
 			queryDone <- struct{}{}
 		}()
 	}
 
-loop:
 	// wait for all the "d" disjoint queries to complete before we return
+	// XXX: Waiting until all queries are done is a vector for DoS attacks:
+	// The disjoint lookup paths that are taken over by adversarial peers
+	// can easily be fooled to go on forever.
+	numQueriesComplete := 0
 	for {
-		select {
-		case <-queryDone:
-			numQueriesComplete++
-			if numQueriesComplete == d {
-				break loop
-			}
-		case <-ctx.Done():
-			break loop
+		<-queryDone
+		numQueriesComplete++
+		if numQueriesComplete == d {
+			break
 		}
 	}
 
-	return queries, nil
+	res := dht.constructLookupResult(queries, targetKadID)
+	return res, nil
 }
 
-// TODO This function should be owned by the DHT as it dosen't really belong to "a query".
-// scorePeerByDistanceAndLatency scores a peer using metrics such as connectendness of the peer, it's distance from the key
-// and it's current known latency.
-func (q query) scorePeerByDistanceAndLatency(p peer.ID, distanceFromKey *big.Int) interface{} {
-	connectedness := q.dht.host.Network().Connectedness(p)
-	latency := q.dht.host.Peerstore().LatencyEWMA(p)
+// constructLookupResult takes the query information and uses it to construct the lookup result
+func (dht *IpfsDHT) constructLookupResult(queries []*query, target kb.ID) *lookupResult {
+	// determine if any queries terminated early
+	completed := true
+	for _, q := range queries {
+		if !(q.isLookupTermination() || q.isStarvationTermination()) {
+			completed = false
+			break
+		}
+	}
 
-	var c int64
-	switch connectedness {
-	case network.Connected:
-		c = 1
-	case network.CanConnect:
-		c = 5
-	case network.CannotConnect:
-		c = 10000
-	default:
-		c = 20
+	// extract the top K not unreachable peers from each query path, as well as their states at the end of the queries
+	var peers []peer.ID
+	peerState := make(map[peer.ID]qpeerset.PeerState)
+	for _, q := range queries {
+		qp := q.queryPeers.GetClosestNotUnreachable(dht.bucketSize)
+		for _, p := range qp {
+			// Since the same peer can be seen in multiple queries use the "best" state for the peer
+			// Note: It's possible that a peer was marked undialable in one path, but wasn't yet tried in another path
+			// for now we're going to return that peer as long as some path does not think it is undialable. This may
+			// change in the future if we track addresses dialed per path.
+			state := q.queryPeers.GetState(p)
+			if currState, ok := peerState[p]; ok {
+				if state > currState {
```

Comment:
This should be fine because the peers are all Reachable as they're derived from `GetClosestNotUnreachable`. However if we'd like more clarity here, I'd be fine with replacing this with `(currState == PeerHeard && (state == PeerWaiting || state == PeerQueried)) || (currState == PeerWaiting && state == PeerQueried)`.

Thoughts?